### ISSUE TITLE
e-sparqlEndpoint: support more query types and more serializations

### ISSUE THREAD

**Issue by jindrichmynarz** [UnifiedViews/Plugins#279]
The current version of e-sparqlEndpoint supports only [`CONSTRUCT`](http://www.w3.org/TR/sparql11-query/#construct) and output in RDFDataUnit. However, SPARQL endpoints support more query types, the most prominent one being [`SELECT`](http://www.w3.org/TR/sparql11-query/#select), and more serializations, such as CSV for `SELECT` results.

Since the DPU currently does not support more query types and serializations, there are cases, in which it complicates pipeline design. For example, I have pipeline that extracts data using `CONSTRUCT` in e-sparqlEndpoint and turns that into CSV via `SELECT` in t-sparqlSelect in the next step because I need the data in CSV for the extractor that follows in the pipeline. This makes pipeline design more complicated that it can be and lay users may have difficulties how to proceed in such cases.

Therefore I would propose to enhance e-sparqlEndpoint with support for `SELECT` and allow users to pick their preferred serialization. Supporting `SELECT` introduces the possibility to have not only RDFDataUnit as the DPU's output, but FilesDataUnit as well. If this would complicate the implementation, we can split the DPU in two: one that produces RDFDataUnit based on `CONSTRUCT` results and second that produces FilesDataUnit based on `SELECT` results. An alternative design is to have a single DPU and both `CONSTRUCT` and `SELECT` produce FilesDataUnit, which can be subsequently converted to RDFDataUnit for `CONSTRUCT` results via t-filesToRdf.

If we decide to enhance the DPU in the proposed way, we should consider how it affects the ongoing work on #241 and if if can be addressed separately after #241 is done.

**Comment by jakubklimek** [comment]
This should be definitely a new DPU, not an enhancement of the existing DPU.

**Comment by jindrichmynarz** [comment]
Could you elaborate why?

**Comment by skrchnavy** [comment]
I agree with @jakubklimek. My view: in common DPU there must be switch to define which output is used (what form of data is to be extracted). And then 2 different configuration dialogs will be available.
So benefit of it is low comparing to independent DPUs.
Similar example: we have  3 sparql transformers.

If we want to enhance existing DPU, then candidate is T-SparqlSelect, where is similar configuration that is requested already done. But question is if we shall have combined Extractors/transformers in one DPU.

**Comment by jindrichmynarz** [comment]
I don't see a reason why 2 different configuration dialogs would be necessary. "Endpoint's URL" and "Query" fields stay the same. An additional "Format" field is added. However, this one needs different values in a select box depending on the type of query. If that is hard to implement, then the approach of having 2 DPUs should be used.

**Comment by skrchnavy** [comment]
It is not hard to implement but benefit for user is not good and is confusing.
2 dialogs are there even on one screen, when you switch output type then you have to change behavior of dialog and also validation functions.
And then user have to use proper output and it is not possible without opening configuration of DPU to know which is active.

**Comment by tomas-knap** [comment]
Yes, two DPUs should be prepared - because if we put everything to one, e-sparqlEndpoint will become more complicated - by having more complicated conf, supporting two outputs (which basically means that  one has to always map outputs to inputs, user may be confused whether CONSTRUCT produces File or RDF data unit, ... SO I would consider this DPU which @jindrichmynarz  is proposing as an alternative to e-sparqlEndpoint.

**Comment by mkl-eea-sk** [comment]
Draft new issues (one  for each new DPU) please. I was able to recognize:
- eSPARQLEnpoint-construct TO FILE data unit (no insertion into internal RDF store is done, results from querying sparql endpoint are saved as files)
- eSPARQLEndpoint-select TO FILE data unit. Does /sparql endpoints support this type of query? do they return .CSV when we issue SELECT query with Accepts: text/csv ?

**Comment by jindrichmynarz** [comment]
To answer your question, format of CSV results for SPARQL `SELECT` query is [standardized](http://www.w3.org/TR/sparql11-results-csv-tsv/), but its support is not required. In practice, the most common implementations (such as Virtuoso or Fuseki) support it.
